src/ngraph/graph_util.hpp

```diff
@@ -78,7 +78,134 @@ namespace ngraph
+    ///        N1[Concat, concatenation_axis=3]   new_N1[Concat, concatenation_axis=3]
+    ///                                                  |                |
+    ///                                                  v                v
+    ///                                               some_user         another_user
+    ///
+    /// (Step 3: N0 and N1 are now dead, nodes will be freed)
+    ///
+    ///    [happens automatically, once all shared_ptrs to N1 are released]
+    ///
+    ///   A----------------------------------------.
+    ///                                            |
+    ///                           B----------------)--.
+    ///                                            |  |
+    ///                                            |  |
+    ///                                      C-----)--)--.
+    ///                                            |  |  |
+    ///                                            v  v  v
+    ///                                           new_N1[Concat, concatenation_axis=3]
+    ///                                                  |                |
+    ///                                                  v                v
+    ///                                               some_user         another_user
+    ///
+    /// NOTE 1: replace_node is not type-safe (the graph is not revalidated).
+    /// For example, the following is allowed, even if node `some_user`
+    /// requires an input of shape 2x2:
+    ///
+    /// (Before)
+    ///      A(shape=2x2)  B(shape=3x3)
+    ///      |
+    ///      v
+    ///   some_user(requires 2x2 input)
+    ///
+    /// (After -- graph is now invalid)
+    ///
+    ///      replace_node(A,B);
+    ///
+    ///      A(shape=2x2)  B(shape=3x3)
+    ///                    |
+    ///                    v
+    ///                 some_user(requires 2x2 input)
+    ///
+    /// NOTE 2: it is possible to insert a cycle into the graph with
+    /// replace_node, resulting in an invalid graph. Care must be taken to
+    /// avoid this. One common example is when you are attempting to insert a
+    /// new node `M` "after"` a node `N`. For example, you might expect this
+    /// to work:
+    ///
+    ///    shared_ptr<Node> M = make_shared<SomeUnaryOp>(N);
+    ///    replace_node(M,N);
+    ///
+    /// The problem is that at replacement time, N itself is a user of M. So
+    /// we end up introducing a cycle as follows:
+    ///
+    ///       N
+    ///       |
+    ///       v
+    ///  other users...
+    ///
+    ///      |||
+    ///      vvv
+    ///
+    ///       N------------>M
+    ///       |
+    ///       v
+    ///  other users...
+    ///
+    ///      |||
+    ///      vvv
+    ///
+    ///               .----.
+    ///              |      |
+    ///              |      |
+    ///       N      `----->M
+    ///                     |
+    ///                     v
+    ///                other users...
```

Comment:
```suggestion
    ///                other users...
    ///
    ///    To avoid the cycle, a valid way to perform the above desired insertion would be,
    ///
    ///        auto new_N = N->copy_with_new_args(N->get_arguments());
    ///        shared_ptr<Node> M = make_shared<SomeUnaryOp>(new_N);
    ///        replace_node(N, M);
```
I verified this works with the below test, 
```
NGRAPH_TEST(${BACKEND_NAME}, replace_node_fanout)
{
    Shape shape{2, 2};
    auto A = make_shared<op::Parameter>(element::f32, shape);
    auto relu0 = make_shared<op::Relu>(A);
    auto relu1 = make_shared<op::Relu>(relu0);
    auto relu2 = make_shared<op::Relu>(relu0);


    auto new_relu = relu0->copy_with_new_args(relu0->get_arguments());
    auto reshape = make_shared<op::Reshape>(new_relu, AxisVector{0, 1}, Shape{4});
    ngraph::replace_node(relu0, reshape);

    auto f = make_shared<Function>(NodeVector{relu1, relu2}, ParameterVector{A});
    auto backend = runtime::Backend::create("${BACKEND_NAME}");

    shared_ptr<runtime::Tensor> a = backend->create_tensor<float>(shape);
    copy_data(a, vector<float>{-1, 2, 3, 4});

    auto handle = backend->compile(f);
    shared_ptr<runtime::Tensor> result1 = backend->create_tensor<float>(shape);
    shared_ptr<runtime::Tensor> result2 = backend->create_tensor<float>(shape);
    backend->call_with_validate(handle, {result1, result2}, {a});


    vector<float> expected{0, 2, 3, 4};
    EXPECT_EQ(expected, read_vector<float>(result1));
    EXPECT_EQ(expected, read_vector<float>(result2));
}
```